### PR TITLE
US4302 and 4303 Pki-Init cache and cacheca option

### PR DESCRIPTION
**Pull request by jim-wang-yutsung**
Implemented pki-init -cache option: generate CA and TLS certs and cache into env. $PKI_CACHE directory in docker container

### CODE REVIEW

File: pkiinit/option/cache.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	return normal, nil
+}
+
+func doCache(eraseCAPrivateKey bool) error {
+	// source
+	generatedDirPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitGeneratedDir)
+
+	if eraseCAPrivateKey {
+		// shreds CA private key before copy to cache dir
+		os.Remove(filepath.Join(generatedDirPath, caServiceName, tlsSecretFileName))
```
Comment (bnevis-i) [inline pkiinit/option/cache.go]: This is not sufficient to securely erase the CA private key.  You need to open the file and do random I/O to overwrite every byte in the file with zeroes, then remove the file.

Comment (bnevis-i) [reply pkiinit/option/cache.go]: You will also want to security erase the CA private key (with -generate) even if you aren't caching the PKI (unless you are explicitly caching the CA).  Or better restated: deploy shouldn't ever deploy the CA private key and it should eventually be shred unless --cacheca is specified.

File: pkiinit/option/optionHelpers.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	pkiCacheDir := os.Getenv(envPkiCache)
+	if pkiCacheDir == "" {
+		return defaultPkiCacheDir
+	}
+	return pkiCacheDir
+}
+
+func deploy(srcDir, destDir string) error {
+	// cleanup the destDir first
+	os.RemoveAll(destDir)
```
Comment (bnevis-i) [inline pkiinit/option/optionHelpers.go]: This (removeall) could be problematic and cause unexpected issues for the end user.

File: pkiinit/option/cache.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	return pkiInitOption == nil || !pkiInitOption.CacheOpt
+}
+
+func cachePkis(cacheca bool) (statusCode exitCode, err error) {
+	// first perform generate func
+	statusCode, err = generatePkis()
+	if err != nil {
+		return statusCode, err
+	}
+
```
Comment (bnevis-i) [inline pkiinit/option/cache.go]: I would lift the destruction of the CA private key out of the cache function and place it here to make it very explicit in the flow.

**Review by bnevis-i — CHANGES_REQUESTED** [review]

File: pkiinit/option/cache.go (hunk 1 above)
Comment (jim-wang-yutsung) [reply pkiinit/option/cache.go]: ok

File: pkiinit/option/cache.go (hunk 3 above)
Comment (jim-wang-yutsung) [reply pkiinit/option/cache.go]: ok

File: pkiinit/option/optionHelpers.go [hunk 4]
```diff
@@ -118,3 +124,32 @@ func getPkiCacheDirEnv() string {
...
+	fileSize := fileInfo.Size()
+	zeroBytes := make([]byte, fileSize)
+	copy(zeroBytes[:], "0")
+
+	// fill the file with zero bytes
+	if _, err := fileHdl.Write(zeroBytes); err != nil {
+		return err
+	}
+
+	return os.Remove(fileToErase)
```
Comment (bnevis-i) [inline pkiinit/option/optionHelpers.go]: Looks like os.Remove might get called before fileHdl.Close().  Probably should close all open file handles before removing the file.

**Review by eno-intel — APPROVED** [review]

File: pkiinit/option/optionHelpers.go [hunk 5]
```diff
@@ -118,3 +124,32 @@ func getPkiCacheDirEnv() string {
...
 }
+
+func deploy(srcDir, destDir string) error {
+	return copyDir(srcDir, destDir)
+}
+
+func secureEraseFile(fileToErase string) error {
+	// grant the file read-write permission first
+	os.Chmod(fileToErase, os.ModePerm)
+	fileHdl, err := os.OpenFile(fileToErase, os.O_RDWR, 0666)
```
Comment (bnevis-i) [inline pkiinit/option/optionHelpers.go]: Is `0666` argument used in this scenario?

**Review by bnevis-i — APPROVED** [review]
LGTM

**Comment by jim-wang-yutsung** [comment]
Waiting for Walt to validate / verify the functionality on his machine. -Jim

**Review by eno-intel — APPROVED** [review]

**Review by bnevis-i — CHANGES_REQUESTED** [review]
I believe there is a bug in the implementation.  If there is a cached pki, it looks like there is a possibility that https://github.com/jim-wang-intel/security-secret-store/blob/3e39367730a86a2cefb54eff70d819006efdba38/pkiinit/option/cache.go#L58 will try to delete a generated CA private key which doesn't exist because no PKI is generated.   the OpenFile will generate an error that is propagated up the call stack.

To reproduce this, it would be necessary to remove the "generated" folder and leave a cached PKI in place, as would happen after a system reboot (please add a unit test for this to make sure it isn't an issue)

**Comment by jim-wang-yutsung** [comment]
>

> 
> 
> I believe there is a bug in the implementation. If there is a cached pki, it looks like there is a possibility that
> 
> https://github.com/jim-wang-intel/security-secret-store/blob/3e39367730a86a2cefb54eff70d819006efdba38/pkiinit/option/cache.go#L58
> will try to delete a generated CA private key which doesn't exist because no PKI is generated. the OpenFile will generate an error that is propagated up the call stack.
> 
> To reproduce this, it would be necessary to remove the "generated" folder and leave a cached PKI in place, as would happen after a system reboot (please add a unit test for this to make sure it isn't an issue)

Good catch, I'll fix it.

**Review by bnevis-i — APPROVED** [review]
LGTM

**Review by bnevis-i — APPROVED** [review]

**Review by eno-intel — APPROVED** [review]

**Pull request merged**